File: 백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<vector<int>> vi;
typedef pair<int, int> pii;
typedef pair<ll, ll> pll;
typedef vector<vector<ll>> vll;
int dx[4] = { 1, -1, 0, 0 };
int dy[4] = { 0, 0, 1, -1 };
int n, m;
int a[10];
bool cache[10];
vector<int> res;

void func(int cnt)
{
    if (cnt == m)
    {
        for (auto to : res)
            cout << to << ' ';
        cout << "\n";
        return;
    }

    for (int i = 1; i <= n; i++)
    {
        if (cache[i] == false)
        {
            cache[i] = true;
            res.push_back(i);
            func(cnt + 1);
            res.pop_back();
            cache[i] = false;
        }
    }
}

int main(void)
{
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);
    cin >> n >> m;
    func(0);
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
n, m = map(int, input().split())
cache = [False] * (n + 1)  # 인덱스 1부터 사용하기 위해 n+1로 설정

def func(cnt, res):
    if cnt == m:
        print(' '.join(map(str, res)))
        return

    for i in range(1, n + 1):
        if not cache[i]:
            cache[i] = True
            res.append(i)
            func(cnt + 1, res)
            res.pop()  # 선택 해제하고 다음 숫자 탐색
            cache[i] = False

func(0, [])  # 초기 카운트 0, 선택된 숫자 리스트는 비어있는 상태에서 시작